### ISSUE TITLE
Clarify that advanced scripts/functions should generally NOT use `$_` / `$PSItem` in their `process` blocks

### ISSUE THREAD

**Issue by mklement0** [MicrosoftDocs/PowerShell-Docs#10944]
### Type of issue

Missing information

### Feedback

Following up from https://github.com/PowerShell/PowerShell/issues/21327:

Of necessity, advanced functions/scripts must explicitly define pipeline-binding parameters.

Inside a `process` block, it is these parameters that should be used rather than `$_` / `$PSItem`, not least because the latter has _no value_ if a given invocation is _argument_- rather than pipeline-based.

(Conversely, non-advanced functions/scripts have no choice but to use `$_` / `$PSItem` to refer to pipeline input.)

To illustrate:

```powershell
function Get-Foo {
  param([Parameter(ValueFromPipeline)] [string] $FooName)
  process {
    "[$FooName] vs. [$_]"
  }
}

Get-Foo -FooName Foo1
```

The above outputs `[Foo1] vs. []`, demonstrating that use of `$_` isn't appropriate for argument-based calls.

### Page URL

https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_functions_advanced_methods?view=powershell-7.4

### Content source URL

https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.4/Microsoft.PowerShell.Core/About/about_Functions_Advanced_Methods.md

### Author

@sdwheeler

### Document Id

c42b871e-8286-5c8e-8dcb-4d8929ab9974

**Comment by sdwheeler** [comment]
The documentation for the [process][1] block says:

> The automatic variable `$_` or `$PSItem` contains the current object in the pipeline for use in the process block. 

Since you are calling the function with a parameter, there is nothing in the pipeline. The function needs to handle the value of the parameter. I think that trying to add an example only confuses the issue.

[1]: https://learn.microsoft.com/powershell/module/microsoft.powershell.core/about/about_functions_advanced_methods#process

**Comment by mklement0** [comment]
That there's nothing in the pipeline is the very point of the issue raised:

* If you declare a pipeline-binding parameter, you get to _choose_ whether to bind it via the pipeline or by argument, and by using the parameter variable rather than `$_`, you don't need to worry about _which method was chosen_.

  * While not the norm, there are cmdlets where by-argument binding _is_ meaningfully supported, namely if the parameter is array-typed, and explicit enumeration is performed inside the `process` block, e.g. `Set-Content foo.txt -Value (1..10)` as the _more efficient alternative to_ `1..10 | Set-Content foo.txt` - that is also what @SetTrend was looking for in https://github.com/PowerShell/PowerShell/issues/21327, and got confused by the current state of the docs.

* However, even in the more typical case (a pipeline-binding parameter such as `-InputObject` that is more of an implementation detail to facilitate pipeline input), _there is never a good reason to use `$_` in the `process` block of an advanced function_ - it is conceptually muddled, and situationally breaks, as demonstrated.

In other words: what causes confusion is the _current_ documentation, by suggesting use of  `$_` in this context.

Please reopen this issue and amend the documentation.

**Comment by mklement0** [comment]
To spell it out in more detail:

* In a _non_-advanced script or function, `$_` / `$PSItem` is the _only_ way to refer to pipeline input in its `process` block. If you tried to explicitly declare a pipeline-binding parameter there, it would _invariably turn into an advanced script/function_, due to the required use of a `[Parameter()]` attribute.

* While you _technically can_ use `$_` / `$PSItem` in the `process` block of _advanced_ scripts/functions too, _you shouldn't_, for the reasons detailed [above](https://github.com/MicrosoftDocs/PowerShell-Docs/issues/10944#issuecomment-1992062041).

**Comment by sdwheeler** [comment]
It's not that I don't understand your point. To me, it's obvious from the documentation that `$_` and `$PSItem` only pertain to pipeline input and if you pass values by parameter then you need to use the parameter variable. What I don't understand is the expectation of the original issue poster.

However, I have made changes that should help.

**Comment by SetTrend** [comment]
Thank you, @sdwheeler, for making the change.

Frankly, I think it's very irritating and not highlighted enough in the documentation that having the user use an _argument_ instead of a _pipeline_ for the _same parameter_ results in _completely different_ behavior, requiring pretty much attention and boilerplate code.

See this simple example:

```powershell
function Test-Pipeline
{
  param
  ( [Parameter(Mandatory = $true, Position = 0, ValueFromPipeline = $true)][int[]]$Argument
  )

  begin
  {
    '/==== Begin ============\'
    "`t`$_: $_, `$PSItem: $PSItem, `$Argument: $Argument"
  }

  process
  {
    ' ---- Process ============'
    "`t`$_: $_, `$PSItem: $PSItem, `$Argument: $Argument"
  }

  end
  {
    '\==== End ============/'
    "`t`$_: $_, `$PSItem: $PSItem, `$Argument: $Argument"
  }
}
```

If the user chooses to use the pipeline, this is the resulting execution workflow:

```powershell
(1..4) | Test-Pipeline
/==== Begin ============\
        $_: , $PSItem: , $Argument:    
 ---- Process ============
        $_: 1, $PSItem: 1, $Argument: 1
 ---- Process ============
        $_: 2, $PSItem: 2, $Argument: 2
 ---- Process ============
        $_: 3, $PSItem: 3, $Argument: 3
 ---- Process ============
        $_: 4, $PSItem: 4, $Argument: 4
\==== End ============/
        $_: 4, $PSItem: 4, $Argument: 4
```

Having the same parameter provided as argument results in a completely different execution workflow and completely different variable values:

```powershell
/==== Begin ============\
        $_: , $PSItem: , $Argument: 1 2 3 4
 ---- Process ============
        $_: , $PSItem: , $Argument: 1 2 3 4
\==== End ============/
        $_: , $PSItem: , $Argument: 1 2 3 4
```

This is not very encouraging …

**Comment by mklement0** [comment]
@SetTrend, indeed, the asymmetry between by-argument vs. by-pipeline multi-value input is unfortunate.
For a previous discussion, see:

* https://github.com/PowerShell/PowerShell/issues/4242
